### PR TITLE
Implement tagging in namespace blob

### LINKED ISSUE #5341: Object Tagging Gaps on Namespace buckets
### Environment info
- Version: **VERSION**
- Deployment: **AZURE | AWS | GCLOUD | ESX | VBOX | DEV**
- Customer: **NAME**

### Actual behavior
GAPS:
A) Namespace s3 as source doesn't support copy object tagging:
1. Since the S3 SDK doesn't return the TagCount in the headObject we cannot know if we should trigger a getObjectTagging in order to get the tags.
2. This works when the source namespace is NooBaa.
3. We can use the header from the readObject in order to know if there are any tags, or we can do the inefficient way of just calling getObjectTagging for every copy object call that is originated from namespace s3.

B) Namespace blob doesn't support object tagging

C) Namespace s3 doesn't return TagCount (we need to take the counter from the headers of read_object_stream and not just the body stream)

### Expected behavior
1. We should be able to populate the tags in the object_sdk.

### Steps to reproduce
1.

### PR DESCRIPTION
**Pull request by nadavMiz**
### Explain the changes
1. implement tagging functionality in namespace blob. implement get_object_tagging, delete_object_tagging, put_object_tagging, and add tags to upload_object. in the process fix problems with uploading tags

### Issues: Fixed #xxx / Gap #xxx
1. partially fixes: https://github.com/noobaa/noobaa-core/issues/5341 (adds tag-count for blob)

### Testing Instructions:
1. 

- [ ] Doc added/updated
- [ ] Tests added

### CODE REVIEW

File: src/sdk/namespace_blob.js [hunk 1]
```diff
@@ -254,13 +254,19 @@ class NamespaceBlob {
...
         let obj;
         const blob_client = this._get_blob_client(params.key);
+        const tagging = params.tagging && Object.fromEntries(params.tagging.map(tag => ([tag.key, tag.value])));
         if (params.copy_source) {
             if (params.copy_source.ranges) {
                 throw new Error('NamespaceBlob.upload_object: copy source range not supported');
             }
-            const sasToken = await this._get_sas_token(params.copy_source.bucket, params.copy_source.key, 'r'); // read permission
+            const tag_permision = params.tagging_copy || params.tagging ? 't' : null;
+            const sasToken = await this._get_sas_token(params.copy_source.bucket, params.copy_source.key, 'r' + tag_permision); // read permission
```
Comment (romayalon) [inline src/sdk/namespace_blob.js:263]: according to node: 
'r' + null = 'rnull'
that's not what you would like to get, you should replace null with ''

File: src/endpoint/s3/s3_utils.js [hunk 2]
```diff
@@ -410,15 +410,15 @@ function parse_tagging_header(req) {
         dbg.error('parse_tagging_header failed', err);
         throw new S3Error(S3Error.MalformedXML);
     }
-    const tag_set = new Set();
+    const tag_map = new Map();
```
[the remark this answers is not in the document]
Comment (nadavMiz) [reply src/endpoint/s3/s3_utils.js:413]: the set only contained tag.key, so the function returned array of keys. the rest of the code expects array of tag objects that contains both key and value. this is wrong regardless of my changes, and caused both s3 and noobaa buckets to fail.
the map is so I can have both the tag object and tag keys separately to check for duplicate keys

**Comment by romayalon** [comment]
@nadavMiz 
Are you familiar with this Azure issue with invalid characters of meta data? https://github.com/noobaa/noobaa-core/pull/7258
I wonder if tags have the same problem, please check that

**Comment by romayalon** [comment]
@nadavMiz 
Also what about some unit tests? :)

**Comment by nadavMiz** [comment]
> @nadavMiz Are you familiar with this Azure issue with invalid characters of meta data? #7258 I wonder if tags have the same problem, please check that



> @nadavMiz Are you familiar with this Azure issue with invalid characters of meta data? #7258 I wonder if tags have the same problem, please check that

as discussed I added a function to check the validity of the tags (_check_valid_tagging). in case the tags are invalid the request is rejected.

**Comment by nadavMiz** [comment]
> @nadavMiz Also what about some unit tests? :)

as discussed there are already unit tests for tags in test_s3_ops. however tags are not supported currently by azurite:
https://github.com/Azure/Azurite#support-matrix
https://github.com/Azure/Azurite/issues/1871 

so for now tested it locally with azure credentials and test passed

however, test "should allow multipart with empty part" fails however since uploading empty parts is not allowed with azure. so I added skip to it in case of azure

**Review by romayalon — APPROVED** [review]

**Pull request merged**